### DISCUSSION TITLE
Debug over SWD

### DISCUSSION THREAD

**Discussion by jarraneil** [micropython/micropython#10705] [RP2040 / Pico]
I am trying to get a programming and debug workflow set up that does not use USB. I am stuck trying to get the REPL over serial.

For SWD I am using a Raspberry Pi4 as my development environment and the tool chain uses openocd to connect to the SWD pins on the RP2040. I have confirmed that this works and can push firmware elf files into the RP2040 and interact with a debugger (gdb-multiarch).

For REPL I have the RP2040 serial port on pins 2 and 3 wired to the Pi4 serial port at /dev/serial0. I have confirmed that this works by running the pico-examples/hello_world/serial application and see the responses in the terminal.

In order to get the REPL going over serial, I modify micropython/ports/rp2/mpconfigport.h with the following one line change
-#define MICROPY_HW_ENABLE_UART_REPL (0) // useful if there is no USB
+#define MICROPY_HW_ENABLE_UART_REPL (1) // useful if there is no USB

The build seems to proceed fine but the load is unresponsive when I push it to the device. If I back out the above change and build again using the same steps the micropython firmware works as expected.

Steps to recreate the issue:
1) checkout latest micropython source code
2) modify one line of code in mpconfigport.h as shown above
3) build micropython rp2
4) push resulting elf file into device using
openocd -f interface/raspberrypi-swd.cfg -f target/rp2040.cfg -c "program build-PICO/firmware.elf verify reset exit"

The same results are obtained if I replace step 4 with drag and drop the newly built uf2 file.

**Comment by jarraneil** [comment]
A bit more information ...  If I run picotool on the unresponsive elf file I get the following i.e. it seems fine but I am not able to connect to it via USB or serial.

picotool info -a ~/micropython/ports/rp2/build-PICO/firmware.elf
File /home/neil/micropython/ports/rp2/build-PICO/firmware.elf:

Program Information
 name:            MicroPython
 version:         v1.19.1-859-g41ed01f13-dirty
 features:        UART REPL
                  USB REPL
                  thread support
 frozen modules:  _boot, _boot_fat, rp2, uasyncio, uasyncio/core,
                  uasyncio/event, uasyncio/funcs, uasyncio/lock,
                  uasyncio/stream, onewire, ds18x20, dht, neopixel
 binary start:    0x10000000
 binary end:      0x1004c6bc
 embedded drive:  0x100a0000-0x10200000 (1408K): MicroPython

Fixed Pin Information
 0:  UART0 TX
 1:  UART0 RX

Build Information
 sdk version:       1.4.0
 pico_board:        pico
 boot2_name:        boot2_w25q080
 build date:        Feb 10 2023
 build attributes:  MinSizeRel

**Reply by th3w** [reply]
I just tested this with a current build "v1.19.1-859-g41ed01f13-dirty on 2023-02-11".
I connected a FT232 USB serial adapter to pin 1 and 2 (GPIO0 and 1) of the Pico_W board.
First I had problems that after flashing, the normal USB Seriell device of the Pico was no longer recognized by Windows. This problem was caused by a double assignment of the UART0. There was still a boot.py on the board, which mirrored the REPL on the UART0 before with os.dupterm().

But I also saw in your first  explanation that you have used pin 2 and 3 to connect the Seriell to USB Adapter to the Pico:
`For REPL I have the RP2040 serial port on pins 2 and 3 wired to the Pi4 serial port at /dev/serial0.`
But according to the PinOut scheme for the PICO and PICO_W board, Pin2 is UART0 RX and Pin3 is GND.

**Reply by robert-hh** [reply]
Maybe @jarraneil talks about GPIO 2 and 3, not about the board pin numbers. That's a constant source of confusion.

**Reply by jarraneil** [reply]
That solved it thank you so much.
And yes I was referring to the RP2040 pin numbers.
I had an old boot.py file in the filesystem that was taking the serial port

**Comment by patfrench** [comment]
Hello,

I have the same problem as you except that I don't have a boot.py file.

#define MICROPY_HW_USB_CDC (0) 
#define MICROPY_HW_ENABLE_UART_REPL (1)

I have a FTDI 3.3V cable on
![image](https://github.com/micropython/micropython/assets/44723412/243065f7-e298-46d4-84e3-4b5f1886c48a)

(with common GND) but I don't have REPL and windows can't find a USB driver. 
![image](https://github.com/micropython/micropython/assets/44723412/fdc5a1ed-cb47-430f-8e44-26e1ddf8dc8e)


1. Which driver should I install ? usbser.sys ?
2. Why don't I have REPL on UART0 ?

thanks in advance

**Reply by projectgus** [reply]
@patfrench The "Board in FS Mode" appears like this because the USB device is still active in the MicroPython firmware, but with USB-CDC disabled it now doesn't have any USB interfaces defined (unless you add some). It's something of a useless device!

You should see the FTDI USB device appear elsewhere in Device Manager, most likely as a COM port. If it's connected correctly then the REPL will be accessible there.

(Having TX and RX swapped by mistake often gets me, FWIW.)

**Reply by patfrench** [reply]
I found (a broken wire between the board and the FTDI cable).
![image](https://github.com/micropython/micropython/assets/44723412/aab8d984-03bf-4d2e-b384-0d09c2a737cf)
thanks aliexpress pffffff

After connecting an oscilloscope to the UART0, I noticed that everything worked.

I'm afraid of polluting PR#9497 with my questions. Can I ask on the PR or on new discuss?

Thanks again

**Reply by projectgus** [reply]
@patfrench A new discussion is a good idea, thanks! If you start one, feel free to @-mention me so I don't miss it.
